README.md
# supersm
Super Symlink Manager. 

![build](https://github.com/peeweep/supersm/workflows/build/badge.svg)
![GitHub last commit](https://img.shields.io/github/last-commit/peeweep/supersm)
![GitHub license](https://img.shields.io/github/license/peeweep/supersm)

### Features

This software looks like [GNU Stow](https://www.gnu.org/software/stow/), but something different:

1. Don't need `--no-folding` anymore.
    Only symlink files, not folder.
2. Don't need `--adopt`anymore.
    If symlink's path already exist file, we will force delete it.

OPTIONS:

```text
Super Symlinks Manager
Allowed options:
  -V [ --version ]                      version message
  -H [ --help ]                         help message
  -T [ --target ] arg                   set target
  -A [ --add ] arg                      Add links
  -D [ --delete ] arg                   Remove links
```

* `--target` will use current path's parent directory if not specified.(absolute path)
* For Windows user: Create symlink on Windows needing to elevate the cmd console as admin.

### Build

Depends: boost-libs

For Debian user:
```
$ sudo apt install libboost-filesystem-dev libboost-program-options-dev
$ cmake -B build && cmake --build build
```

For macOS user:
```
$ brew install boost
$ cmake -B build && cmake --build build
```

For Windows user:

Install [msys2](https://www.msys2.org/), then add the absolute path of `msys64\mingw64\bin` and `msys64\mingw64\include` to PATH.

```
$ pacman -S mingw-w64-x86_64-boost mingw-w64-x86_64-cmake mingw-w64-x86_64-gcc
$ cmake -G "MinGW Makefiles" -B build
$ cmake --build build
```

For Arch user:

[AUR](https://aur.archlinux.org/packages/supersm)


debian/changelog
supersm (0.5.2) unstable; urgency=medium

  * Disable static linking.

 -- peeweep <peeweep@0x0.ee>  Thu, 01 Jul 2021 01:50:43 +0800

supersm (0.5.1) unstable; urgency=medium

  * Initial Release.

 -- peeweep <peeweep@0x0.ee>  Tue, 01 Jun 2021 16:59:13 +0800


src/file.h
//
// Created by peeweep on 2019-12-21.
//

#ifndef SUPERSM_FILE_H
#define SUPERSM_FILE_H

#include <boost/filesystem.hpp>
#include <iostream>

class file {
 public:
  static std::string split_path(std::string source,
                                std::string project_split_string) {
    return source.erase(0, project_split_string.size());
  }

  static void create_dir_if_not_exist(boost::filesystem::path path) {
    if ((boost::filesystem::exists(path) and
         !boost::filesystem::is_directory(path)) or
        boost::filesystem::is_symlink(path)) {
      std::cout << path << " is a file, delete it and create folder"
                << std::endl;
      boost::filesystem::remove(path);
      boost::filesystem::create_directories(path);
    } else if (!boost::filesystem::exists(path)) {
      std::cout << path << " does not exist, create it" << std::endl;
      boost::filesystem::create_directories(path);
    }
  }

  static void remove_file_if_already_exist(boost::filesystem::path path) {
    if (boost::filesystem::exists(path) or boost::filesystem::is_symlink(path))
      boost::filesystem::remove(path);
  }

  static void symlink_all_files(const boost::filesystem::path &project,
                                const std::string &project_split_string,
                                const std::string &target_string) {
    for (auto &iter : boost::filesystem::directory_iterator(project))
      if (boost::filesystem::is_directory(iter))
        symlink_all_files(iter, project_split_string, target_string);
      else {
        boost::filesystem::path symlink_path =
            boost::filesystem::path(target_string) /
            boost::filesystem::path(
                split_path(iter.path().string(), project_split_string));
        create_dir_if_not_exist(symlink_path.parent_path());
        remove_file_if_already_exist(symlink_path);
        boost::filesystem::create_symlink(iter, symlink_path);
        std::cout << "ln -sf " << iter.path() << " " << symlink_path
                  << std::endl;
      }
  }

  static void is_symlink_path_exist(const boost::filesystem::path &symlink_path,
                                    bool &run_next_loop) {
    if (!boost::filesystem::is_symlink(symlink_path)) {
      std::string type;
      std::cout
          << symlink_path
          << " is not a symlink! Are you sure continue run next loop? Y/N? "
             "(default:Y) ";
      std::getline(std::cin, type);
      if (type == "y" or type == "Y" or type.empty()) {
        run_next_loop = true;
        return;
      } else if (type == "n" or type == "N") {
        std::cout << "Let's Quit" << std::endl;
        exit(1);
      } else {
        std::cout << "wrong value, please re-type" << std::endl;
        is_symlink_path_exist(symlink_path, run_next_loop);
      }
    }
  }

  static void remove_all_symlinks(const boost::filesystem::path &project,
                                  const std::string &project_split_string,
                                  const std::string &target_string) {
    for (auto &iter : boost::filesystem::directory_iterator(project))
      if (boost::filesystem::is_directory(iter))
        remove_all_symlinks(iter, project_split_string, target_string);
      else {
        boost::filesystem::path symlink_path =
            boost::filesystem::path(target_string) /
            boost::filesystem::path(
                split_path(iter.path().string(), project_split_string));

        bool run_next_loop = false;
        is_symlink_path_exist(symlink_path, run_next_loop);
        if (run_next_loop)
          continue;

        if (boost::filesystem::read_symlink(symlink_path).string() !=
            iter.path()) {
          std::cout << symlink_path
                    << " already is a symlink but not the same as "
                    << boost::filesystem::read_symlink(symlink_path).string()
                    << std::endl;
          exit(1);
        }
        boost::filesystem::remove(symlink_path);
        std::cout << "Delete symlink: " << iter.path() << std::endl;
      }
  }
};

#endif  // SUPERSM_FILE_H


src/links.h
//
// Created by peeweep on 2020-03-06.
//

#ifndef SUPERSM_LINKS_H
#define SUPERSM_LINKS_H

#include <boost/filesystem.hpp>
#include <iostream>

namespace fs = boost::filesystem;

class links {
 public:
  links(std::vector<fs::path> v_source_directory, fs::path target_directory) {
    this->v_source_directory = v_source_directory;
    this->target_directory   = target_directory;
  }

  void links_all_files(fs::path        source_directory,
                       const fs::path& source_directory_saved) {
    for (auto& iter : fs::directory_iterator(source_directory)) {
      if (fs::is_directory(iter)) {
        links_all_files(iter, source_directory_saved);
      } else {
        // iter is source_file_path
        // target_directory: ~/target_directory
        // source_file_path: ~/source_directory/folder1/file1
        // get the right path: /folder1/file1
        // after split: ~/b/folder1/file1
        auto symlink_file =
            fs::path(this->target_directory.string() +
                     std::string(iter.path().string())
                         .erase(0, source_directory_saved.string().size()));

        // create parent directory if not exist
        if (!fs::exists(symlink_file.parent_path()))
          fs::create_directories(symlink_file.parent_path());

        // remove old file, create new symlink
        if (fs::exists(symlink_file) or fs::is_symlink(symlink_file)) {
          fs::remove(symlink_file);
        }
        fs::create_symlink(iter, symlink_file);

        std::cout << "ln -sf " << iter << " " << symlink_file << std::endl;
      }
    }
  }

  void add_links() {
    for (auto& it : this->v_source_directory) {
      if (!fs::is_directory(it)) {
        std::cerr << "Project " << it << " does not exist!\n";
        exit(1);
      }
      fs::path source_directory = fs::absolute(it);
      links_all_files(source_directory, source_directory);
    }
  }

  void del_all_links(fs::path        source_directory,
                     const fs::path& source_directory_saved) {
    for (auto& iter : fs::directory_iterator(source_directory)) {
      if (fs::is_directory(iter)) {
        del_all_links(iter, source_directory_saved);
      } else {
        // iter is source_file_path
        // target_directory: ~/target_directory
        // source_file_path: ~/source_directory/folder1/file1
        // get the right path: /folder1/file1
        // after split: ~/b/folder1/file1
        auto symlink_file =
            fs::path(this->target_directory.string() +
                     std::string(iter.path().string())
                         .erase(0, source_directory_saved.string().size()));
        // remove file if already exist, and it isn't a folder
        if (fs::is_symlink(symlink_file) or
            (!fs::is_directory(symlink_file) and fs::exists(symlink_file))) {
          fs::remove(symlink_file);
          std::cout << "delete symlink " << symlink_file << std::endl;
        } else {
          std::cout << symlink_file << "is not a file, skip" << std::endl;
        }
      }
    }
  }

  void del_links() {
    for (auto& it : this->v_source_directory) {
      if (!fs::is_directory(it)) {
        std::cerr << "Project " << it << " does not exist!\n";
        exit(1);
      }
      fs::path source_directory = fs::absolute(it);
      del_all_links(source_directory, source_directory);
    }
  }

 private:
  std::vector<fs::path> v_source_directory;
  fs::path              target_directory;
};

#endif  // SUPERSM_LINKS_H


src/common.h
//
// Created by peeweep on 2019-12-18.
//

#ifndef SUPERSM_COMMON_H
#define SUPERSM_COMMON_H

#define VERSION 0.4;

#include <boost/filesystem.hpp>
#include <boost/program_options.hpp>
#include <iostream>

#include "file.h"
class common {
 public:
  static void print_help_and_exit(
      const boost::program_options::options_description& description) {
    std::cout << "Super Symlinks Manager. [supersm] Version " << VERSION;
    std::cout << std::endl;
    std::cout << description;
    exit(1);
  }

  static boost::program_options::variables_map process_options(
      int argc, char const* argv[]) {
    // add options
    boost::program_options::options_description desc("Allowed options");
    desc.add_options()("version,v", "version message")("help,h",
                                                       "help message")(
        "supersm,s",
        boost::program_options::value<std::vector<boost::filesystem::path>>()
            ->multitoken(),
        "supersm project")(
        "undo,u",
        boost::program_options::value<std::vector<boost::filesystem::path>>()
            ->multitoken(),
        "undo supersm")(
        "target,t",
        boost::program_options::value<boost::filesystem::path>()->default_value(
            boost::filesystem::current_path().parent_path()),
        "set target");

    // store map
    boost::program_options::variables_map variablesMap;

    // set as supersm if no options
    boost::program_options::positional_options_description p;
    p.add("supersm", -1);
    boost::program_options::store(
        boost::program_options::command_line_parser(argc, argv)
            .options(desc)
            .allow_unregistered()
            .positional(p)
            .run(),
        variablesMap);
    boost::program_options::notify(variablesMap);

    // print version and exit
    if (variablesMap.count("version")) {
      std::cout << VERSION;
      std::cout << std::endl;
      exit(1);
    }

    // print help and exit
    if (variablesMap.count("help") or argc <= 1
        //    or variablesMap["supersm"].empty()
    )
      print_help_and_exit(desc);

    boost::filesystem::path target_abs_path = boost::filesystem::absolute(
        variablesMap["target"].as<boost::filesystem::path>());

    if (!variablesMap["supersm"].empty())
      // range loop supersm project
      for (auto& it :
           variablesMap["supersm"].as<std::vector<boost::filesystem::path>>()) {
        const boost::filesystem::path& project_path = it;
        if (!boost::filesystem::is_directory(project_path)) {
          std::cout << "supersm project " << project_path << " Don't Exist!"
                    << std::endl;
          exit(1);
        }
        boost::filesystem::path project_abs_path =
            boost::filesystem::absolute(project_path);
        file::symlink_all_files(project_abs_path,
                                project_abs_path.string() + "/",
                                target_abs_path.string());
      }

    if (!variablesMap["undo"].empty())
      // range loop remove supersm project
      for (auto& it :
           variablesMap["undo"].as<std::vector<boost::filesystem::path>>()) {
        boost::filesystem::path project_abs_path =
            boost::filesystem::absolute(it);
        file::remove_all_symlinks(project_abs_path,
                                  project_abs_path.string() + "/",
                                  target_abs_path.string());
      }

    return variablesMap;
  }
};

#endif  // SUPERSM_COMMON_H


src/core/core.cpp
#include <boost/program_options.hpp>
#include <filesystem>
#include <iostream>
#define VERSION 0.1

void printFile(const std::filesystem::path &path) {
  for (auto &it : std::filesystem::directory_iterator(path)) {
    if (std::filesystem::is_directory(it))
      printFile(it);
    else
      std::cout << std::filesystem::path(it).string() << std::endl;
  }
}

void print_help_and_exit(
    boost::program_options::options_description description) {
  std::cout << "Super Symlinks Manager. [SuperSM] Version " << VERSION
            << std::endl;
  std::cout << description;
  exit(1);
}

int main(int argc, char const *argv[]) {
  std::string default_target = "DIR";

  // add options
  boost::program_options::options_description desc("Allowed options");
  desc.add_options()("version,v", "version message")("help,h", "help message")(
      "target,t",
      boost::program_options::value<std::filesystem::path>()->default_value(
          default_target),
      "set target");

  // store map
  boost::program_options::variables_map variablesMap;
  boost::program_options::store(
      boost::program_options::parse_command_line(argc, argv, desc),
      variablesMap);
  boost::program_options::notify(variablesMap);

  // set as target if no options
  boost::program_options::positional_options_description p;
  p.add("target", -1);
  boost::program_options::store(
      boost::program_options::command_line_parser(argc, argv)
          .options(desc)
          .positional(p)
          .run(),
      variablesMap);

  // print version and exit
  if (variablesMap.count("version")) {
    std::cout << VERSION << std::endl;
    exit(1);
  }

  // print help and exit
  if (variablesMap.count("help") or argc <= 1)
    print_help_and_exit(desc);

  // print target and exit
  if (variablesMap["target"].as<std::filesystem::path>().string() !=
      default_target) {
    std::cout << "Target was set to "
              << variablesMap["target"].as<std::filesystem::path>()
              << std::endl;
    exit(1);
  }

  return 0;
}


src/options.h
//
// Created by peeweep on 2020-03-06.
//

#ifndef SUPERSM_OPTIONS_H
#define SUPERSM_OPTIONS_H
#define SUPERSM_VERSION_MAJOR 0
#define SUPERSM_VERSION_MINOR 5
#define SUPERSM_VERSION_REVISION 2

#define SUPERSM_AUX_STR_EXP(__A) #__A
#define SUPERSM_AUX_STR(__A) SUPERSM_AUX_STR_EXP(__A)

#define SUPERSM_VERSION_STR              \
  SUPERSM_AUX_STR(SUPERSM_VERSION_MAJOR) \
  "." SUPERSM_AUX_STR(SUPERSM_VERSION_MINOR) "." SUPERSM_AUX_STR(SUPERSM_VERSION_REVISION) " "

#include <boost/filesystem.hpp>
#include <boost/program_options.hpp>
#include <iostream>

#include "links.h"

namespace fs = boost::filesystem;
namespace po = boost::program_options;

class options {
 public:
  static void parse_option(int argc, char const* argv[]) {
    // add options
    po::options_description desc("Allowed options");
    desc.add_options()
        // version message
        ("version,V", "version message")
        // help message
        ("help,H", "help message")
        // target option's default value is current_path's parent_path
        ("target,T",
         po::value<fs::path>()->default_value(fs::current_path().parent_path()),
         "set target")
        // Add links
        ("add,A", po::value<std::vector<fs::path>>()->multitoken(), "Add links")
        // Remove links
        ("delete,D", po::value<std::vector<fs::path>>()->multitoken(),
         "Remove links");

    // store map
    po::variables_map variablesMap;
    // set as supersm if no options
    boost::program_options::positional_options_description p;
    p.add("add", -1);
    po::store(po::command_line_parser(argc, argv)
                  .options(desc)
                  .allow_unregistered()
                  .positional(p)
                  .run(),
              variablesMap);
    po::notify(variablesMap);

    // Convert "target" option's value to absolute path
    fs::path target = fs::absolute(variablesMap["target"].as<fs::path>());

    // print version message
    if (!variablesMap["version"].empty()) {
      std::cout << SUPERSM_VERSION_STR << std::endl;
      exit(1);
    }
    // print help
    if (!variablesMap["help"].empty() or argc <= 1) {
      std::cout << "Super Symlinks Manager.";
      std::cout << "[supersm] Version " << SUPERSM_VERSION_STR << std::endl;
      std::cout << desc;
      exit(1);
    }
    // Add links
    if (!variablesMap["add"].empty()) {
      auto link =
          links(variablesMap["add"].as<std::vector<fs::path>>(), target);
      link.add_links();
    }
    // remove links
    if (!variablesMap["delete"].empty()) {
      auto link =
          links(variablesMap["delete"].as<std::vector<fs::path>>(), target);
      link.del_links();
    }
  }
};

#endif  // SUPERSM_COMMON_H
